src/openms_gui/source/VISUAL/Spectrum2DCanvas.cpp

```diff
@@ -326,100 +326,85 @@ namespace OpenMS
+        } 
+        std::sort(n_s.begin(), n_s.end());
+        n_peaks_in_scan = n_s[1]; // median
       }
-
-      // determine spacing for whole data
-      double min_spacing_mz = 1.0;
-      double average_spacing_rt = 1.0;
-      Size n_scans = peak_map.size();
-      {
-        vector<float> mz_spacing;
-        for (Size i = 0; i != n_scans; ++i)
-        {
-          // skip non MS1 and empty spectra
-          Size ms_level = peak_map[i].getMSLevel();
-          Size n_peaks = peak_map[i].size();
-          if (ms_level != 1 || n_peaks < 2)
-          {
-            continue;
-          }
-          double current_mz_spacing =  (peak_map[i][n_peaks - 1].getMZ() - peak_map[i][0].getMZ()) / n_peaks;
-          mz_spacing.push_back(current_mz_spacing);
-        }
-        sort(mz_spacing.begin(), mz_spacing.end());
-        min_spacing_mz = !mz_spacing.empty() ? mz_spacing[0] : 1.0;
+      
+      double ratio_data2pixel_rt = n_ms1_scans / (double)rt_pixel_count;
+      double ratio_data2pixel_mz = n_peaks_in_scan / (double)mz_pixel_count;
 
 #ifdef DEBUG_TOPPVIEW
-        cout << "BEGIN " << __PRETTY_FUNCTION__ << endl;
-        cout << "min spacing mz:" << min_spacing_mz << endl;
+      std::cerr << rt_min << ":" << rt_max <<"   " << mz_min << ":" << mz_max << "\n";
+      std::cerr << n_ms1_scans << "rt " << n_peaks_in_scan << "ms\n";
+      std::cerr << rt_pixel_count << " x " << mz_pixel_count << " px\n";
+      std::cerr << ratio_data2pixel_rt << " " << ratio_data2pixel_mz << " ratio\n";
 #endif
+      // minimum fraction of image expected to be filled with data
+      // if not reached, we upscale point size
+      const double MIN_COVERAGE = 0.2;
 
-        {
-          vector<float> rts;
-          for (Size i = 0; i != n_scans; ++i)
-          {
-            // skip non MS1 and empty spectra
-            Size ms_level = peak_map[i].getMSLevel();
-            Size n_peaks = peak_map[i].size();
-            if (ms_level != 1 || n_peaks == 0)
-            {
-              continue;
-            }
-            rts.push_back(peak_map[i].getRT());
-          }
-          sort(rts.begin(), rts.end());
-          if (rts.size() > 2)
-          {
-            average_spacing_rt = (rts[rts.size() - 1] - rts[0]) / (double)rts.size();
-          }
-        }
-      }
+      bool has_low_pixel_coverage = ratio_data2pixel_rt < MIN_COVERAGE || ratio_data2pixel_mz < MIN_COVERAGE;
 
-      // Determine whether several peaks are expected to be drawn on the same pixel
-      if (n_peaks_in_middle_scan > mz_pixel_count || n_ms1_scans > rt_pixel_count)
+      // Are several peaks expected to be drawn on the same pixel in either RT or m/z?
+      // --> thin out and show only maxima
+      // Also, we cannot upscale in this mode (since we operate on the buffer directly, i.e. '1 data point == 1 pixel'
+      if (!has_low_pixel_coverage & (n_peaks_in_scan > mz_pixel_count || n_ms1_scans > rt_pixel_count))
```

Comment:
Ouch. Yes.
-- 
Sent via mobile phone.

Am 6. April 2015 22:36:06 MESZ, schrieb Hannes Roest <notifications@github.com>:
>>  
>> -      // Determine whether several peaks are expected to be drawn on
>the same pixel
>> -      if (n_peaks_in_middle_scan > mz_pixel_count || n_ms1_scans >
>rt_pixel_count)
>> +      // Are several peaks expected to be drawn on the same pixel in
>either RT or m/z?
>> +      // --> thin out and show only maxima
>> +      // Also, we cannot upscale in this mode (since we operate on
>the buffer directly, i.e. '1 data point == 1 pixel'
>> +      if (!has_low_pixel_coverage & (n_peaks_in_scan >
>mz_pixel_count || n_ms1_scans > rt_pixel_count))
>
>looks like a bug to me? did you mean `&&` here? 
>
>PS: caught by cdash
>http://cdash.openms.de/testDetails.php?test=24132385&build=125852
>
>---
>Reply to this email directly or view it on GitHub:
>https://github.com/OpenMS/OpenMS/pull/1336/files#r27827739